### PR TITLE
Convert result of multivariate polynomial evaluation into correct parent

### LINKED ISSUE #33373: Incorrect parent when evaluating constant multivariate polynomial
<div id="comment:0"></div>

When evaluating a univariate polynomial, it correctly takes the parent of the input into account:

```
sage: k.<a> = GF(2^4)
sage: R.<x> = PolynomialRing(k)
sage: f = R(1)
sage: S.<y> = PolynomialRing(k)
sage: f(y).parent()
Univariate Polynomial Ring in y over Finite Field in a of size 2^4
```
Multivariate polynomials do not:

```
sage: k.<a> = GF(2^4)
sage: R.<x> = PolynomialRing(k, 1)
sage: f = R(1)
sage: S.<y> = PolynomialRing(k, 1)
sage: f(y).parent()
Finite Field in a of size 2^4
```

Component: **algebra**

Keywords: **pAdicBordeaux**

Author: **David Roe**

Reviewer: **Marc Mezzarobba**

_Issue created by migration from https://trac.sagemath.org/ticket/33373_

### PR DESCRIPTION
**Pull request by mezzarobba**
Importing trac/u/roed/incorrect_parent_when_evaluating_constant_multivariate_polynomial...

(meant to fix #33373)

### CODE REVIEW

**Comment by mezzarobba** [comment]
I already reviewed @roed314's commit (on trac, see the linked issue), so we only need someone to take a look at mine.

File: src/sage/modular/modform_hecketriangle/graded_ring_element.py [hunk 1]
```diff
@@ -1580,7 +1582,7 @@ def _q_expansion_cached(self, prec, fix_d, subs_d, d_num_prec, fix_prec = False)
         d = self.parent().get_d(fix_d, d_num_prec)
         q = self.parent().get_q(prec, fix_d, d_num_prec)
 
-        qexp = sum([(qexp.coefficients()[m]).subs(d=d) * q**qexp.exponents()[m] for m in range(len(qexp.coefficients()))])
+        qexp = sum([((qexp.coefficients()[m]).subs(d=d)) * q**qexp.exponents()[m] for m in range(len(qexp.coefficients()))])
```
Comment (tscrim) [inline src/sage/modular/modform_hecketriangle/graded_ring_element.py]: This extra set of parentheses seems unless (and IMO, makes it harder to parse).

File: src/sage/rings/polynomial/multi_polynomial_libsingular.pyx [hunk 2]
```diff
@@ -2075,29 +2084,28 @@ cdef class MPolynomial_libsingular(MPolynomial):
...
-        cdef poly *res    # ownership will be transferred to us in the next line
-        singular_polynomial_call(&res, self._poly, _ring, coerced_x, MPolynomial_libsingular_get_element)
-        res_parent = coercion_model.common_parent(parent._base, *x)
-
-        if res == NULL:
-            return res_parent(0)
-        if p_LmIsConstant(res, _ring):
-            sage_res = si2sa( p_GetCoeff(res, _ring), _ring, parent._base )
-            p_Delete(&res, _ring)            # sage_res contains copy
+                sage_res += c*mul([ x[i]**m[i] for i in m.nonzero_positions()])
```
Comment (tscrim) [inline src/sage/rings/polynomial/multi_polynomial_libsingular.pyx]:
```suggestion
                sage_res += c * mul([x[i] ** m[i] for i in m.nonzero_positions()])
```
PEP8 stuff

File: src/sage/rings/polynomial/multi_polynomial_libsingular.pyx [hunk 3]
```diff
@@ -2075,29 +2084,28 @@ cdef class MPolynomial_libsingular(MPolynomial):
...
+            if res == NULL:
+                return res_parent(0)
+            if p_LmIsConstant(res, _ring):
+                sage_res = si2sa( p_GetCoeff(res, _ring), _ring, parent._base )
+                p_Delete(&res, _ring)            # sage_res contains copy
+            else:
+                sage_res = new_MP(parent, res)   # pass on ownership of res to sage_res
 
-        if parent(sage_res) is not res_parent:
+        if sage_res.parent() is not res_parent:
```
Comment (tscrim) [inline src/sage/rings/polynomial/multi_polynomial_libsingular.pyx]: This is actually slightly slower. Are we also sure that `sage_res` is an `Element`? If so, we can do
```suggestion
        if sage_res._parent is not res_parent:
```

File: src/sage/schemes/riemann_surfaces/riemann_surface.py [hunk 4]
```diff
@@ -2193,7 +2194,7 @@ def local_N(ct, rt):
                 z_1 = a0lc.abs() * prod((cz - r).abs() - rho_z for r in a0roots)
                 n = minpoly.degree(CCzg.gen(1))
                 ai_new = [
-                    (minpoly.coefficient({CCzg.gen(1): i}))(z=cz + self._CCz.gen(0))
+                    (CCz(minpoly.coefficient({CCzg.gen(1): i})))(z=cz + self._CCz.gen(0))
```
Comment (tscrim) [inline src/sage/schemes/riemann_surfaces/riemann_surface.py]:
```suggestion
                    CCz(minpoly.coefficient({CCzg.gen(1): i}))(z=cz + self._CCz.gen(0))
```
Are these parentheses [an elegant weapon for a more civilized age](https://xkcd.com/297/)?

File: src/sage/schemes/toric/points.py [hunk 5]
```diff
@@ -849,7 +849,7 @@ def inhomogeneous_equations(self, ring, nonzero_coordinates, cokernel):
         z = [ring.zero()] * nrays
         for i, value in zip(nonzero_coordinates, z_nonzero):
             z[i] = value
-        return [poly(z) for poly in self.polynomials]
+        return [poly.change_ring(ring)(z) for poly in self.polynomials]
```
Comment (tscrim) [inline src/sage/schemes/toric/points.py:852]:
```suggestion
        newpoly = poly.change_ring(ring)
        return [newpoly(z) for poly in self.polynomials]
```
So we do not create a new polynomial for each item in the list.

**Review by tscrim — CHANGES_REQUESTED** [review]
Just a few more tweaks, but LGTM otherwise.

File: src/sage/modular/modform_hecketriangle/graded_ring_element.py (hunk 1 above)
Comment (mezzarobba) [reply src/sage/modular/modform_hecketriangle/graded_ring_element.py]: I agree; I think I committed this line by accident

File: src/sage/rings/polynomial/multi_polynomial_libsingular.pyx (hunk 2 above)
Comment (mezzarobba) [reply src/sage/rings/polynomial/multi_polynomial_libsingular.pyx]: fine with me

File: src/sage/schemes/riemann_surfaces/riemann_surface.py (hunk 4 above)
Comment (mezzarobba) [reply src/sage/schemes/riemann_surfaces/riemann_surface.py]: Here I don't know what the original author's intention was, but I'm fine with removing the extra parentheses.

File: src/sage/schemes/toric/points.py (hunk 5 above)
Comment (mezzarobba) [reply src/sage/schemes/toric/points.py:852]: uh, the comprehension is over `poly`

File: src/sage/rings/polynomial/multi_polynomial_libsingular.pyx (hunk 3 above)
Comment (mezzarobba) [reply src/sage/rings/polynomial/multi_polynomial_libsingular.pyx]: ok

File: src/sage/schemes/toric/points.py (hunk 5 above)
Comment (tscrim) [reply src/sage/schemes/toric/points.py:852]: Right, duh. I was thinking the loop was over `z`.

**Review by tscrim — APPROVED** [review]
Thanks. LGTM. I think the failure in `ell_field.py` is unrelated to this ticket.

**Comment by mezzarobba** [comment]
Thanks for the review!

**Pull request merged**